### PR TITLE
[Draft] Rework the "Overview of the Compiler" chapter

### PR DESCRIPTION
**Pull request by jieyouxu**
Context: https://github.com/rust-lang/rustc-dev-guide/issues/674

## Reworking the "Overview of the Compiler" chapter

I've noticed that the current Overview chapter seems to have several issues that make it difficult for new contributors to understand the compiler's structure and architecture at a high-level:

- Mixed detail levels.
- Some areas are very briefly mentioned or omitted.
- Still very hard to actually get a reasonably full high-level picture of the compiler.

See also discussion during a rustc-dev-guide reading club session (cf. <https://rust-lang.zulipchat.com/#narrow/stream/196385-t-compiler.2Fwg-rustc-dev-guide/topic/.5Breading.20club.5D.202023.2E02.2E26.20chapters.2022.20to.2030/near/423423778>).

This PR aims to rework the Overview of the Compiler chapter to make it more friendly for new contributors and be more comprehensive and to address some of these shortcomings.

The R0 draft here is intentionally incomplete: I worked on this draft in a separate HackMD document, but it's at a point where it has sufficient content to allow people to start giving feedback, suggestions and contributions (and I also want to start fixing the links and reworking diagrams so they can be embedded in a mdBook-compatible fashion).

Some inline remarks (presented as quote blocks) are left in during draft phase, will be removed when ready.

[Rendered text](https://github.com/jieyouxu/rustc-dev-guide/blob/overview-rework/src/overview/overview.md)

## Revision History

Revision history is maintained through individual commits, so they are not repeated here.

### CODE REVIEW

File: src/overview/overview.md [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+
+See [The Typed High-level Intermediate
+Representation](https://rustc-dev-guide.rust-lang.org/thir.html) for more
+details.
+
+#### HIR Body Type-checking
+
+> 🚧 **TODO** 🚧 (Remove the cat meme when complete with draft)
+
+![](https://hackmd.io/_uploads/ry2CM6gAa.jpg)
```
Comment (Nadrieril) [inline src/overview/overview.md:379]: :rofl: plz don't remove the cat meme

File: src/overview/overview.md [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+> - Motivate the area's existence -- what does it do to / with the inputs? Why?
+> - Introduce the outputs. Briefly motivate why the outputs are what they are.
+> - Want to know more? Link to details chapter here. Link to "Additional
+>   References" section below.
+
+> :warning: **Where does trait system come in?** / **Where do we describe the
+> trait system here?** AFAIK, it's both dependent and and provides information
+> to multiple areas. Link: <https://github.com/rust-lang/rust/pull/121848>
+
+#### Static-Single Assignment (SSA) Code Generation
```
Comment (Nadrieril) [inline src/overview/overview.md:206]: You're missing a section about codegen. Idk if it's querified or not

Comment (jieyouxu) [reply src/overview/overview.md:206]: ~~We don't talk about codegen~~ Yeah I'm not completely sure about codegen atm.

File: src/overview/overview.md [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+>
+> We _embrace_ the query architecture and talk about areas _backwards_ compared
+> to traditional compiler texts. Since information dependence of the various
+> `rustc` areas form a Direct Acyclic Graph (DAG) at the query view (not acyclic
+> at the areas view), it makes sense to start at the root -- the final
+> executable -- and traverse the graph in a breadth-first traversal, all the way
+> until we get to name resolution. At that point, we can zoom out and talk about
+> the non-querified areas (in either reverse order like the querified areas, or
+> forwards order like traditional passes-based compilers, depending on which
+> order feels more natural).
```
Comment (Nadrieril) [inline src/overview/overview.md:13]: I'm not super convinced by this. I find it easier to follow the flow of information in the parse->codegen direction. That way I can know what information is available at every stage. Similarly, whether an area is querified or not feels like an implementation detail; it doesn't affect the general understanding of which information is computed where.

I do agree that it's important for contributors to understand the querified nature of rustc. I think the walkthrough chapter would be a good place to walk through the "real" execution order, starting from the cli invocation all the way to parsing the files.

File: src/overview/overview.md [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+  satisfy the `where` clauses on the ADT.
+
+The notion of well-formedness can be inductively extended for entities that
+contain types, such as trait predicates. A trait predicate `T0: Trait<P1, ...>`
+is well-formed if `T0` and `P1, ...` are well-formed in addition to `Trait`'s
+`where` clauses being satisfied.
+
+HIR well-formedness checking is an early collection of checks on mostly surface
+Rust syntax (such as parameters, local variables and return types) which are run
+before further type-checking and inference, for two purposes:
```
Comment (Nadrieril) [inline src/overview/overview.md:424]:
```suggestion
The HIR is _well-formed_ (WF) when it satisfies a set of requirements that include:
- Well-formedness of types in type signatures
- Object safety
- Function call ABI requirements (e.g. for "rust-call")
- Check if receiver type can be used as a self type
- Variances check
- etc
These checks are run before further type-checking and inference, for two purposes:
```
(I just took the list from your Zulip post, I don't know if that's a sensible list)

File: src/overview/overview.md (hunk 3 above)
Comment (jieyouxu) [reply src/overview/overview.md:13]: Yeah, I could shuffle the organization around the "usual" way without it being *too* disruptive. I'll see what people think w.r.t. ordering.

File: src/overview/overview.md (hunk 1 above)
Comment (jieyouxu) [reply src/overview/overview.md]: something something retaining the cat meme might require an FCP in the final version /jk

File: src/overview/overview.md (hunk 4 above)
Comment (jieyouxu) [reply src/overview/overview.md]: I added these examples to extend the notion of HIR wfcheck to cover not just type wfcheck.

File: src/overview/overview.md (hunk 3 above)
Comment (jieyouxu) [reply src/overview/overview.md]: I flipped the summary ordering to go from parse->codegen. Thanks for the feedback.

File: src/overview/overview.md (hunk 2 above)
Comment (jieyouxu) [reply src/overview/overview.md]: Stubbed out a section on the 3 codegen backends

**Comment by Noratrieb** [comment]
@jieyouxu meow (do you plan to continue working on this at some point)

**Comment by jieyouxu** [comment]
> @jieyouxu meow (do you plan to continue working on this at some point)

yes meow

**Comment by jieyouxu** [comment]
I'm closing this PR because I think I have a better idea for how to organize the overview and I didn't really vibe with this organization (of information dependency flow, I still find it confusing myself). I want to talk about it in the reading club and w/ t-compiler ppl ~next week. (And it's easier to start from a fresh plate.)

**Pull request closed without merging**